plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,403 @@
+import hu.bme.mit.gamma.expression.model.QuantifierExpression;
+import hu.bme.mit.gamma.expression.model.RationalLiteralExpression;
+import hu.bme.mit.gamma.expression.model.RationalTypeDefinition;
+import hu.bme.mit.gamma.expression.model.RecordAccessExpression;
+import hu.bme.mit.gamma.expression.model.RecordLiteralExpression;
+import hu.bme.mit.gamma.expression.model.RecordTypeDefinition;
+import hu.bme.mit.gamma.expression.model.ReferenceExpression;
+import hu.bme.mit.gamma.expression.model.SelectExpression;
+import hu.bme.mit.gamma.expression.model.SubtractExpression;
+import hu.bme.mit.gamma.expression.model.Type;
+import hu.bme.mit.gamma.expression.model.TypeDefinition;
+import hu.bme.mit.gamma.expression.model.TypeReference;
+import hu.bme.mit.gamma.expression.model.UnaryExpression;
+import hu.bme.mit.gamma.expression.model.UnaryMinusExpression;
+import hu.bme.mit.gamma.expression.model.UnaryPlusExpression;
+import hu.bme.mit.gamma.expression.model.VoidTypeDefinition;
+import hu.bme.mit.gamma.util.GammaEcoreUtil;
+
+public class ExpressionTypeDeterminator2 {
+	public static final ExpressionTypeDeterminator2 INSTANCE = new ExpressionTypeDeterminator2();
+	protected ExpressionTypeDeterminator2() {}
+	
+	protected final GammaEcoreUtil ecoreUtil = GammaEcoreUtil.INSTANCE;
+	protected final ExpressionUtil expressionUtil = ExpressionUtil.INSTANCE;
+	protected final ExpressionModelFactory factory = ExpressionModelFactory.eINSTANCE;
+	
+	public Type getType(Expression expression) {
+		if (expression instanceof Declaration) {
+			return ExpressionModelDerivedFeatures.getTypeDefinition((Declaration)expression);
+		}
+		if (expression instanceof TypeReference) {
+			Type type = getType(expression);
+			for (TypeReference reference : ecoreUtil.getAllContentsOfType(type, TypeReference.class)) {
+				TypeDefinition typeDefinition = ExpressionModelDerivedFeatures.getTypeDefinition(reference);
+				ecoreUtil.replace(typeDefinition, reference);
+			}
+			return ExpressionModelDerivedFeatures.getTypeDefinition(type);
+		}
+		if (expression instanceof BooleanLiteralExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof IntegerLiteralExpression) {
+			return factory.createIntegerTypeDefinition();
+		}
+		if (expression instanceof RationalLiteralExpression) {
+			return factory.createRationalTypeDefinition();
+		}
+		if (expression instanceof DecimalLiteralExpression) {
+			return factory.createDecimalTypeDefinition();
+		}
+		if (expression instanceof EnumerationLiteralExpression) {
+			return factory.createEnumerationTypeDefinition();
+		}
+		if (expression instanceof IntegerRangeLiteralExpression) {
+			return factory.createIntegerRangeTypeDefinition();
+		}
+		if (expression instanceof RecordLiteralExpression) {
+			return factory.createRecordTypeDefinition();
+		}
+		if (expression instanceof ArrayLiteralExpression) {
+			return factory.createArrayTypeDefinition();
```

Comment:
Here we have to specify the element type of the array (setElementType(...)). For this, we have to evaluate the type of elements stored in the array literal (e.g., first element).